validator/sawtooth_validator/journal/chain.py

```diff
@@ -543,11 +587,42 @@ def on_block_validated(self, commit_new_block, result):
                                                   result["uncommitted_batches"]
                                                   )
 
+                    # Remove the block from the pending blocks queue and
+                    # verify any pending blocks.
+                    self._verify_descendants(self._chain_head)
+
+                # If the block was determine to be invalid.
+                elif new_block.status == BlockStatus.Invalid:
+                    # Since the block is invalid, we will never accept any
+                    # blocks that are descendants of this block.  We are going
+                    # to go through the pending blocks and remove all
+                    # descendants we find and mark the corresponding block
+                    # as invalid.
                     pending_blocks = \
-                        self._blocks_pending.pop(
-                            self._chain_head.block.header_signature, [])
-                    for pending_block in pending_blocks:
-                        self._verify_block(pending_block)
+                        self._blocks_pending.pop(new_block.identifier, [])
+                    while pending_blocks:
+                        pending_block = pending_blocks.pop()
+                        pending_block.status = BlockStatus.Invalid
+
+                        LOGGER.debug(
+                            'Marking descendant block invalid: %s',
+                            pending_block)
+
+                        pending_blocks.extend(
+                            self._blocks_pending.pop(
+                                pending_block.identifier,
+                                []))
+
+                # The block is otherwise valid, but we have determined we
+                # don't want it as the chain head.
+                else:
+                    LOGGER.info('Rejected new chain head: %s', new_block)
+
+                    # Remove the block from the pending blocks and verify
+                    # any immediate descendants that arrived while we were
+                    # processing it.
+                    self._verify_descendants(new_block)
```

Comment:
Calling this `_verify_descentants` really seems to hide what the code is doing.  The fact that there is a comment above the call (here and in the other places it's called), describing what the function is doing, tells us that this is a badly named function.  

Since it is called by itself here at this line, the expectation is that if it failed verification (the implied idea by using "verify" - there are situations where this could fail), it should thrown an exception.  Looking at the implementation, it returns false, but it has the side-effect of removing a block from the pending queue.  We're still not really looking at the result of return value in this instance.

The old name didn't hide this, but perhaps the reason it seems wrong is it is doing to many things.  This could be divided into two operations:

```
remaining_blocks = \
            self._blocks_pending.pop(new_block.identifier, [])
self._verify_all(remaining_blocks)
```
where verify_all does the "verify" portion of the operation.   It takes the descendants out of the equation, as well, and opens the function up to broader usage.  

Part of the confusion about this function name, for me anyway, may also be related to the overloading of the word "verify" that some members of the team think about in it's domain-specific way, and I'm still thinking about the use of "verify" in terms of an general operation .